src/cluster/backend/zeromq/ZeroMQ.cc

```diff
@@ -0,0 +1,537 @@
+        {.socket = log_pull, .name = "log_pull", .handler = HandleLogMessages},
+    };
+
+    std::vector<zmq::pollitem_t> poll_items(sockets.size());
+
+    while ( true ) {
+        for ( size_t i = 0; i < sockets.size(); i++ )
+            poll_items[i] = {.socket = sockets[i].socket.handle(), .fd = 0, .events = ZMQ_POLLIN | ZMQ_POLLERR};
+
+        // Awkward.
+        std::vector<std::vector<MultipartMessage>> rcv_messages(sockets.size());
+        try {
+            int r = zmq::poll(poll_items, std::chrono::seconds(-1));
+            ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: r=%d", r);
+
+            for ( size_t i = 0; i < poll_items.size(); i++ ) {
+                const auto& item = poll_items[i];
+                ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: items[%lu]=%s %s %s\n", i, sockets[i].name.c_str(),
+                                           item.revents & ZMQ_POLLIN ? "pollin " : "",
+                                           item.revents & ZMQ_POLLERR ? "err" : "");
+
+                if ( item.revents & ZMQ_POLLERR ) {
+                    // What should we be doing? Re-open sockets? Terminate?
+                    ZEROMQ_THREAD_PRINTF("poll: error: POLLERR on socket %zu %s %p revents=%x\n", i,
+                                         sockets[i].name.c_str(), item.socket, item.revents);
+                }
+
+                // Nothing to do?
+                if ( (item.revents & ZMQ_POLLIN) == 0 )
+                    continue;
+
+                bool consumed_one = false;
+
+                // Read messages from the socket.
+                do {
+                    zmq::message_t msg;
+                    rcv_messages[i].emplace_back(); // make room for a multipart message
+                    auto& into = rcv_messages[i].back();
+
+                    // Only receive up to poll_max_messages from an individual
+                    // socket. Move on to the next when exceeded. The last pushed
+                    // message (empty) is popped at the end of the loop.
+                    if ( poll_max_messages > 0 && rcv_messages[i].size() > poll_max_messages ) {
+                        ZEROMQ_DEBUG_THREAD_PRINTF(DebugFlag::POLL, "poll: %s rcv_messages[%zu] full!\n",
+                                                   sockets[i].name.c_str(), i);
+                        break;
+                    }
+
+                    consumed_one = false;
+                    bool more = false;
+
+                    // Read a multi-part message.
+                    do {
+                        auto recv_result = sockets[i].socket.recv(msg, zmq::recv_flags::dontwait);
+                        if ( recv_result ) {
+                            consumed_one = true;
+                            more = msg.more();
+                            into.emplace_back(std::move(msg));
+                        }
+                        else {
+                            // EAGAIN and more flag set? Try again!
```

Comment:
It's a little tricky to follow the semantics here (though admittedly I haven't looked up that `recv()`!). Can it happen that `recv_result` is false but there's more to read, and `more` isn't true yet? Might be worth a debug message below the `continue`.

Reply:
Let me stare at this again, I'm not at all happy with how it looks either.

The upside is that for multipart messages there's a guarantee that either we get all of them. So once more=true, we're guaranteed to receive all the other parts.  `recv_result` only holds the size of individual parts, so I don't think what you describe is possible.

It doesn't make the code better to read :-(